File: src/mbio/packages/tool_lab/seq_reverse.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Getopt::Long;
use File::Basename qw(basename dirname);
use Data::Dumper;
my ($help,$input,$type,$output);

#######################################################################################

# ------------------------------------------------------------------
# GetOptions
# ------------------------------------------------------------------
GetOptions(
				"help|?" =>\&USAGE,
				"input:s"=>\$input,
				"type:s"=>\$type,
				"output:s"=>\$output,
				) or &USAGE;
&USAGE unless ($input and $output);
$input = ABSOLUTE_DIR($input);
$output = ABSOLUTE_DIR($output);
$type ||= "rc";
#######################################################################################
open REF,$input;
open OUT,">>$output/output.fasta";
my (%RAWfasta,%NEWfasta,$seq);
while (<REF>) {
	chomp;
	next if ($_ eq "" || /^$/);
	if($_ =~ /\>/){
		$RAWfasta{$_} = "";
		$seq = $_;
	}else{
		$RAWfasta{$seq} .= $_;
	}
}
close REF;

foreach  my $name (keys %RAWfasta) {
	$NEWfasta{$name} = $RAWfasta{$name};
	if($type eq "r"){
		$NEWfasta{$name} = reverse $RAWfasta{$name};
	}elsif($type eq "c"){
		$NEWfasta{$name} = $RAWfasta{$name};
		$NEWfasta{$name} =~ tr/ATCGatcg/TAGCtagc/;
	}elsif($type eq "rc"){
		$NEWfasta{$name} = reverse $RAWfasta{$name};
		$NEWfasta{$name} =~ tr/ATCGatcg/TAGCtagc/;
	}else{
		print "Please provide the right type!";
	}
	print OUT "$name\n$NEWfasta{$name}\n";
}

close OUT;

sub ABSOLUTE_DIR #$pavfile=&ABSOLUTE_DIR($pavfile);
{
	my $cur_dir=`pwd`;chomp($cur_dir);
	my ($in)=@_;
	my $return="";
	if(-f $in){
		my $dir=dirname($in);
		my $file=basename($in);
		chdir $dir;$dir=`pwd`;chomp $dir;
		$return="$dir/$file";
	}elsif(-d $in){
		chdir $in;$return=`pwd`;chomp $return;
	}else{
		warn "Warning just for file and dir \n$in";
		exit;
	}
	chdir $cur_dir;
	return $return;
}


sub USAGE{#
	my $usage =<<"USAGE";
Usage :
	-help	help information
	-input	<file> input fasta file
	-type	<string>	r: reverse, c:complement, rc:reverse and complement
	-output	<file>	output file
USAGE
	print $usage;
	exit;
}

File: src/mbio/packages/tool_lab/seq_reverse.py
#!usr/bin/python
import sys
 
file=open(sys.argv[1],'r')
op=sys.argv[2]
name=sys.argv[3]
atcg=""
first=''
ot=''
outname=''
def complement(s):
	ns=''
        for i in s:
                if i=='A' or i=='a':
                        ns=ns+'T'
                elif i=='T' or i=='t':
                        ns=ns+'A'
                elif i=='C' or i=='c':
                        ns=ns+'G'
                elif i=='G' or i=='g':
                        ns=ns+'C'
                elif i=='N' or i=='n':
                        ns=ns+'N'
        return ns

for i in file:
        i=i.strip("\n")
        if i.find(">",0,2)==-1 :
                atcg=atcg+i     
        else:
                first=i 
if op == 'r':
	ot=atcg[::-1]
	outname=name+".r.fasta"
elif op == 'c':
	ot=complement(atcg)
	outname=name+".c.fasta"
elif op == 'rc':
	ot1=atcg[::-1]
	ot=complement(ot1)
	outname=name+".rc.fasta"
#elif op == 'o':
#	ot=atcg
#	outname=name+".oneline.fna"
else:
	print 'input wrong\n'

otf=open(outname,'w')
otf.write(first+"\n"+ot+'\n')
otf.close()
file.close()
